File: fixture-gen/src/cmds/ergoscript/eval/coll_exists.rs
//! Phase 2f Coll HOFs Task 9 — Exists eval fixtures (fourth lambda HOF).
//!
//! Short-circuits on first `true`; returns `false` for empty input.
//!
//! Cost: Mixed pattern — outer charged on FULL input length BEFORE the loop;
//! per-iter charged only for VISITED items.
//!
//! Sigma-rust ref: `ergotree-interpreter/src/eval/coll_exists.rs:12-69`
//!   line 18: let input_v = self.input.eval(env, ctx)?;
//!   line 19: let condition_v = self.condition.eval(env, ctx)?;
//!   line 29: ctx.add_jit_cost(5)?;                          // per-iter (inside closure)
//!   line 46-58: elem_tpe check: coll.elem_tpe() != &*self.elem_tpe
//!   line 60: ctx.add_per_item_jit_cost(3, 1, 10, n)?;       // outer (BEFORE loop, FULL n)
//!   line 62-66: for item in normalized_input_vals { ... if res { return true; } }
//!
//! CRITICAL: Outer cost at line 60 is AFTER children are evaluated and AFTER the
//! elem_tpe check, but BEFORE the for-item loop at line 62. It charges the FULL
//! input length regardless of how many items are visited due to short-circuiting.
//!
//! Outer cost formula: `add_per_item_jit_cost(base=3, per_chunk=1, chunk_size=10, n)`:
//!   cost = base + per_chunk * ceil(n / chunk_size)
//!   For n=0:    3 + 1 * ceil(0/10)    = 3 + 0   = 3
//!   For n=3:    3 + 1 * ceil(3/10)    = 3 + 1   = 4
//!   For n=12:   3 + 1 * ceil(12/10)   = 3 + 2   = 5
//!   For n=1000: 3 + 1 * ceil(1000/10) = 3 + 100 = 103
//!
//! Smoking-gun: n=1000, short-circuit at item 1.
//!   Outer = addPerItemCost(3, 1, 10, 1000) = 103 (FULL length).
//!   Per-iter = 1 * 5 = 5 (only item 1 visited).
//!   Arm contribution = 103 + 5 = 108 (plus child eval costs).
//!   This proves the outer charges FULL n, not the visited count.
//!
//! EXISTS opcode = LAST_CONSTANT_CODE(112) + new_op_code(62) = 174 = 0xAE
//!
//! Fixture entries (10 total):
//!   1. coll_exists_happy                     — [1,2,3].exists(x => x > 2)  → Boolean(true) (3 visited)
//!   2. coll_exists_sg_full_outer_cost         — [true,false,false,...x1000].exists(x => x) → Boolean(true) (1 visited)
//!      Smoking-gun: outer=103, per-iter=5, arm=108. NOT outer(n=1)+5=9.
//!   3. coll_exists_no_match                  — [1,2,3].exists(x => x > 10) → Boolean(false) (3 visited)
//!   4. coll_exists_empty                     — [].exists(_ => true)         → Boolean(false) (0 visited)
//!   5. coll_exists_sg_n12                    — n=12, all false, proves outer chunking
//!   6. coll_exists_elem_tpe_mismatch         — input SInt, declared elem SLong → 'coll-elem-tpe-mismatch'
//!   7. coll_exists_lambda_not_callable       — condition is non-Lambda → 'lambda-not-callable'
//!   8. coll_exists_lambda_result_type_mismatch — body returns Int → 'lambda-result-type-mismatch'
//!   9. coll_exists_not_coll                  — input is SInt → 'coll-input-not-coll'
//!  10. coll_exists_cost_limit                — jitCostLimit too low → 'cost-limit-exceeded'

use ergotree_interpreter::eval::test_util::try_eval_out;
use ergotree_ir::chain::context::Context;
use ergotree_ir::ergo_tree::{ErgoTree, ErgoTreeHeader};
use ergotree_ir::mir::bin_op::{BinOp, BinOpKind, RelationOp};
use ergotree_ir::mir::coll_exists::Exists;
use ergotree_ir::mir::if_op::If;
use ergotree_ir::mir::expr::Expr;
use ergotree_ir::mir::func_value::{FuncArg, FuncValue};
use ergotree_ir::mir::val_def::ValId;
use ergotree_ir::mir::val_use::ValUse;
use ergotree_ir::mir::value::Value;
use ergotree_ir::serialization::SigmaSerializable;
use ergotree_ir::types::stype::SType;
use serde::Serialize;
use serde_json::{json, Value as JsonValue};
use sigma_test_util::force_any_val;

use super::common::value_to_json;

#[derive(Serialize)]
pub struct CollExistsFixture {
    pub name: String,
    pub tree_bytes_hex: String,
    pub opts_json: JsonValue,
    /// null for error entries
    pub expected_value_json: JsonValue,
    /// 0 for error entries
    pub expected_cost: u64,
    /// null for success entries
    pub expected_error_code: JsonValue,
}

#[derive(Serialize)]
pub struct CollExistsFixtureFile {
    pub corpus: &'static str,
    pub entries: Vec<CollExistsFixture>,
}

/// Build an Exists expr: `input.exists(x => body)` where `x` is bound to `ValId(1)`.
/// `x_tpe` is the element type of `input` (= condition lambda's arg type).
/// Body must have type SBoolean.
fn build_exists(input: Expr, x_tpe: SType, body: Expr) -> anyhow::Result<Expr> {
    let condition: Expr = FuncValue::new(
        vec![FuncArg { idx: ValId(1), tpe: x_tpe }],
        body,
    )
    .into();
    Ok(Exists::new(input, condition)?.into())
}

fn build_tree(expr: Expr) -> anyhow::Result<(ErgoTree, String)> {
    let tree = ErgoTree::new(ErgoTreeHeader::v0(false), &expr)?;
    let hex = hex::encode(tree.sigma_serialize_bytes()?);
    Ok((tree, hex))
}

fn success_entry(name: &str, expr: Expr) -> anyhow::Result<CollExistsFixture> {
    let (tree, hex) = build_tree(expr)?;
    let ctx = force_any_val::<Context>();
    let val: Value<'static> = try_eval_out(&tree.proposition()?, &ctx)?;
    Ok(CollExistsFixture {
        name: name.into(),
        tree_bytes_hex: hex,
        opts_json: json!({}),
        expected_value_json: value_to_json(&val),
        expected_cost: ctx.jit_cost_value(),
        expected_error_code: json!(null),
    })
}

fn cost_limit_entry(name: &str, expr: Expr, limit: u64) -> anyhow::Result<CollExistsFixture> {
    let (_tree, hex) = build_tree(expr)?;
    Ok(CollExistsFixture {
        name: name.into(),
        tree_bytes_hex: hex,
        opts_json: json!({ "jitCostLimit": limit }),
        expected_value_json: json!(null),
        expected_cost: 0,
        expected_error_code: json!("cost-limit-exceeded"),
    })
}

pub fn generate() -> anyhow::Result<CollExistsFixtureFile> {
    let mut entries = Vec::new();

    // ── 1. coll_exists_happy ──────────────────────────────────────────────────
    // [1,2,3].exists(x => x > 2) → Boolean(true)
    // Matches at item 3 (x=3 > 2); all 3 items visited (items 1 and 2 are false).
    // x: SInt; body = x > 2 (BinOp GT)
    {
        let coll: Expr = Expr::Const(vec![1i32, 2i32, 3i32].into());
        let body: Expr = BinOp {
            kind: BinOpKind::Relation(RelationOp::Gt),
            left: Box::new(Expr::ValUse(ValUse { val_id: ValId(1), tpe: SType::SInt })),
            right: Box::new(Expr::Const(2i32.into())),
        }
        .into();
        let expr = build_exists(coll, SType::SInt, body)?;
        entries.push(success_entry("coll_exists_happy", expr)?);
    }

    // ── 2. coll_exists_sg_full_outer_cost ────────────────────────────────────
    // SMOKING-GUN: [true, false, false, ..., false].exists(x => x) — n=1000, match at item 1.
    //
    // Critical subtlety: outer cost at sigma-rust coll_exists.rs:60 is charged BEFORE
    // the for-loop (line 62) using the FULL input length, even though short-circuit
    // exits after item 1.
    //
    // Expected arm contribution:
    //   outer = addPerItemCost(3, 1, 10, 1000) = 3 + ceil(1000/10)*1 = 3 + 100 = 103
    //   per-iter = 1 * 5 = 5  (only item 1 visited before short-circuit)
    //   arm total = 108
    //
    // This is NOT outer(3, 1, 10, 1) + 5 = 4 + 5 = 9 (wrong — would happen if outer
    // charged only visited items).
    //
    // x: SBoolean; body = x (ValUse directly, returns the boolean element)
    // Input: [true, false, false, ..., false] — 1000 items, first is true.
    {
        let mut items = vec![false; 1000];
        items[0] = true; // first item triggers true → short-circuit
        let coll: Expr = Expr::Const(items.into());
        // body = x (the element itself, type SBoolean)
        let body: Expr = Expr::ValUse(ValUse { val_id: ValId(1), tpe: SType::SBoolean });
        let expr = build_exists(coll, SType::SBoolean, body)?;
        entries.push(success_entry("coll_exists_sg_full_outer_cost", expr)?);
    }

    // ── 3. coll_exists_no_match ───────────────────────────────────────────────
    // [1,2,3].exists(x => x > 10) → Boolean(false)
    // No item matches; all 3 items visited (no short-circuit).
    {
        let coll: Expr = Expr::Const(vec![1i32, 2i32, 3i32].into());
        let body: Expr = BinOp {
            kind: BinOpKind::Relation(RelationOp::Gt),
            left: Box::new(Expr::ValUse(ValUse { val_id: ValId(1), tpe: SType::SInt })),
            right: Box::new(Expr::Const(10i32.into())),
        }
        .into();
        let expr = build_exists(coll, SType::SInt, body)?;
        entries.push(success_entry("coll_exists_no_match", expr)?);
    }

    // ── 4. coll_exists_empty ──────────────────────────────────────────────────
    // [].exists(_ => true) → Boolean(false)
    // Empty input — loop never runs. Per-iter = 0. Outer = addPerItemCost(3, 1, 10, 0) = 3.
    // Returns false (sigma-rust coll_exists.rs:68: Ok(false.into())).
    {
        let coll: Expr = Expr::Const(Vec::<bool>::new().into());
        let body: Expr = Expr::Const(true.into());
        let expr = build_exists(coll, SType::SBoolean, body)?;
        entries.push(success_entry("coll_exists_empty", expr)?);
    }

    // ── 5. coll_exists_sg_n12 ────────────────────────────────────────────────
    // n=12, all false — smoking-gun for chunked outer cost.
    // n=3:  outer = 3 + ceil(3/10)*1  = 3 + 1 = 4
    // n=12: outer = 3 + ceil(12/10)*1 = 3 + 2 = 5
    // Proves outer changes at chunk boundary (no short-circuit: all 12 visited).
    {
        let items = vec![false; 12];
        let coll: Expr = Expr::Const(items.into());
        let body: Expr = Expr::ValUse(ValUse { val_id: ValId(1), tpe: SType::SBoolean });
        let expr = build_exists(coll, SType::SBoolean, body)?;
        entries.push(success_entry("coll_exists_sg_n12", expr)?);
    }

    // ── 6. coll_exists_elem_tpe_mismatch ─────────────────────────────────────
    // Coll[Int].exists but Exists's declared elem_tpe = SLong → 'coll-elem-tpe-mismatch'
    //
    // Sigma-rust coll_exists.rs:46-52: `if coll.elem_tpe() != &*self.elem_tpe → EvalError`.
    // Exists::new derives elem_tpe from input SColl, so we must construct Exists directly.
    //
    // We build Exists with:
    //   - input: Coll[Int] (runtime elem = SInt)
    //   - elem_tpe: SLong (mismatches SInt → triggers the check)
    //   - condition: valid FuncValue with SLong arg (for TS: condition.args[0].tpe = SLong)
    //
    // We do NOT call sigma-rust try_eval_out — it would return EvalError.
    {
        let coll: Expr = Expr::Const(vec![1i32, 2i32].into()); // Coll[Int] at runtime
        // Condition: (x: SLong) => true — declared arg is SLong, but runtime elem is SInt
        let condition: Expr = FuncValue::new(
            vec![FuncArg { idx: ValId(1), tpe: SType::SLong }],
            Expr::Const(true.into()),
        )
        .into();
        // Construct Exists directly with mismatched elem_tpe (SLong vs runtime SInt)
        use std::sync::Arc;
        let exists_expr: Expr = Exists {
            input: Box::new(coll),
            condition: Box::new(condition),
            elem_tpe: Arc::new(SType::SLong),
        }
        .into();
        let tree = ErgoTree::new(ErgoTreeHeader::v0(false), &exists_expr)?;
        let hex = hex::encode(tree.sigma_serialize_bytes()?);

        entries.push(CollExistsFixture {
            name: "coll_exists_elem_tpe_mismatch".into(),
            tree_bytes_hex: hex,
            opts_json: json!({}),
            expected_value_json: json!(null),
            expected_cost: 0,
            expected_error_code: json!("coll-elem-tpe-mismatch"),
        });
    }

    // ── 7. coll_exists_lambda_not_callable ───────────────────────────────────
    // Exists(Coll[Int], Const(SBoolean, true)) → 'lambda-not-callable'
    //
    // Condition is a Const (not a FuncValue) so evaluating it yields a Boolean SValue
    // (not a Lambda), which extractFuncValue rejects.
    //
    // Exists::new requires condition.tpe() == SFunc — Const(true) has type SBoolean.
    // Use raw bytes (EXISTS opcode = 0xAE) to bypass.
    //
    // Raw layout:
    //   [0x00]  ErgoTreeHeader v0(false)
    //   [0xAE]  EXISTS opcode = 112 + 62 = 174 = 0xAE
    //   [coll_bytes]   Const(Coll[Int], [1,2,3])
    //   [const_bytes]  Const(SBoolean, true)
    //
    // We do NOT call sigma-rust try_eval_out — it would error.
    {
        let coll: Expr = Expr::Const(vec![1i32, 2i32, 3i32].into());
        let const_cond: Expr = Expr::Const(true.into());

        let coll_bytes = coll.sigma_serialize_bytes()?;
        let const_bytes = const_cond.sigma_serialize_bytes()?;

        let mut tree_bytes = Vec::new();
        tree_bytes.push(0x00u8); // ErgoTreeHeader v0(false)
        tree_bytes.push(0xAEu8); // EXISTS opcode = 112 + 62 = 174 = 0xAE
        tree_bytes.extend_from_slice(&coll_bytes);
        tree_bytes.extend_from_slice(&const_bytes);

        entries.push(CollExistsFixture {
            name: "coll_exists_lambda_not_callable".into(),
            tree_bytes_hex: hex::encode(&tree_bytes),
            opts_json: json!({}),
            expected_value_json: json!(null),
            expected_cost: 0,
            expected_error_code: json!("lambda-not-callable"),
        });
    }

    // ── 8. coll_exists_lambda_result_type_mismatch ───────────────────────────
    // [1, 2].exists(x => if (x == 1) { true } else { x }) → 'lambda-result-type-mismatch'
    //
    // The condition body is `If(x==1, true, x)`. Its compile-time type (true-branch type)
    // is SBoolean (If::tpe returns true-branch type → Const(true).tpe() → SBoolean).
    // FuncValue::new computes t_range = body.tpe() = SBoolean.
    //
    // At eval time:
    //   - item x=1: condition true  → returns Boolean(true) → OK → SHORT-CIRCUIT → true
    //
    // Wait — if x=1 returns true, Exists short-circuits immediately with true.
    // We never reach the type mismatch for x=2.
    //
    // Instead, use [2, 1] so first item causes the mismatch:
    //   - item x=2: condition false → returns Int(2) → type SInt ≠ SBoolean → THROWS
    //   (we never reach x=1)
    //
    // TS: Exists body must return Boolean. Runtime mismatch on x=2 triggers
    // 'lambda-result-type-mismatch'.
    //
    // Exists::new validates t_range == SBoolean (If's tpe is SBoolean), so the
    // standard build path accepts this. We do NOT call sigma-rust try_eval_out —
    // it would error because try_extract_into::<bool>() on Int(2) fails.
    {
        let coll: Expr = Expr::Const(vec![2i32, 1i32].into()); // Coll[Int] — 2 first!
        let x_use: Expr = Expr::ValUse(ValUse { val_id: ValId(1), tpe: SType::SInt });
        // condition: x == 1
        let cond_expr: Expr = BinOp {
            kind: BinOpKind::Relation(RelationOp::Eq),
            left: Box::new(x_use.clone()),
            right: Box::new(Expr::Const(1i32.into())),
        }
        .into();
        // body: if (x == 1) { true } else { x }
        // If::tpe() = true_branch.tpe() = SBoolean
        // false_branch = x (SInt at runtime) — type mismatch on x=2!
        let body: Expr = If {
            condition: Box::new(cond_expr),
            true_branch: Box::new(Expr::Const(true.into())),
            false_branch: Box::new(x_use),
        }
        .into();
        let expr = build_exists(coll, SType::SInt, body)?;
        let (_tree, hex) = build_tree(expr)?;

        entries.push(CollExistsFixture {
            name: "coll_exists_lambda_result_type_mismatch".into(),
            tree_bytes_hex: hex,
            opts_json: json!({}),
            expected_value_json: json!(null),
            expected_cost: 0,
            expected_error_code: json!("lambda-result-type-mismatch"),
        });
    }

    // ── 9. coll_exists_not_coll ───────────────────────────────────────────────
    // Exists(Const(SInt, 42), cond) → 'coll-input-not-coll'
    //
    // `Exists::new` validates that input is SColl — bypass via raw bytes.
    // EXISTS opcode = LAST_CONSTANT_CODE(112) + new_op_code(62) = 174 = 0xAE
    //
    // Raw layout:
    //   [0x00]  ErgoTreeHeader v0(false)
    //   [0xAE]  EXISTS opcode
    //   [int_const_bytes]    Const(SInt, 42)
    //   [condition_bytes]    FuncValue((x: SInt) => true)
    {
        let int_const: Expr = Expr::Const(42i32.into());
        let condition: Expr = FuncValue::new(
            vec![FuncArg { idx: ValId(1), tpe: SType::SInt }],
            Expr::Const(true.into()),
        )
        .into();

        let const_bytes = int_const.sigma_serialize_bytes()?;
        let cond_bytes = condition.sigma_serialize_bytes()?;

        let mut tree_bytes = Vec::new();
        tree_bytes.push(0x00u8); // ErgoTreeHeader v0(false) → 0x00
        tree_bytes.push(0xAEu8); // EXISTS opcode = 112 + 62 = 174 = 0xAE
        tree_bytes.extend_from_slice(&const_bytes);
        tree_bytes.extend_from_slice(&cond_bytes);

        entries.push(CollExistsFixture {
            name: "coll_exists_not_coll".into(),
            tree_bytes_hex: hex::encode(&tree_bytes),
            opts_json: json!({}),
            expected_value_json: json!(null),
            expected_cost: 0,
            expected_error_code: json!("coll-input-not-coll"),
        });
    }

    // ── 10. coll_exists_cost_limit ────────────────────────────────────────────
    // jitCostLimit too low → cost-limit-exceeded before eval completes.
    // Limit = 1 guarantees the very first charge overflows.
    {
        let coll: Expr = Expr::Const(vec![1i32, 2i32, 3i32].into());
        let body: Expr = Expr::Const(true.into());
        let expr = build_exists(coll, SType::SInt, body)?;
        entries.push(cost_limit_entry("coll_exists_cost_limit", expr, 1)?);
    }

    Ok(CollExistsFixtureFile {
        corpus: "eval_coll_exists",
        entries,
    })
}

File: packages/ergoscript/src/eval/coll-exists.ts
/**
 * Exists evaluator arm — fourth lambda HOF (phase 2f Coll HOFs Task 9).
 *
 * Returns true if at least one element of a collection satisfies a boolean
 * predicate (lambda). Short-circuits on the first true element; returns false
 * for empty collections or when no element matches.
 *
 * Cost: Mixed pattern — outer + per-iter.
 *   Sigma-rust ref: ergotree-interpreter/src/eval/coll_exists.rs:12-69
 *
 *   Outer (line 60, AFTER input/condition eval AND elem_tpe check, BEFORE loop):
 *     ctx.add_per_item_jit_cost(3, 1, 10, n)  where n = FULL inputColl.length
 *     NOTE: outer cost is (3, 1, 10) — same as Fold, DIFFERENT from Map/Filter's (20, 1, 10).
 *     CRITICAL: n is the FULL input length, charged BEFORE the loop. Short-circuiting
 *     at item 1 of 1000 still incurs outer = addPerItemCost(3, 1, 10, 1000) = 103.
 *
 *   Per-iter (line 29, inside closure, BEFORE body eval):
 *     ctx.add_jit_cost(5) — only for VISITED items (short-circuit reduces visits).
 *
 * Eval order (sigma-rust coll_exists.rs:18-68):
 *   1. Eval input  → must be Coll  (throws 'coll-input-not-coll')
 *   2. Eval condition → must be Lambda (throws 'lambda-not-callable')
 *   3. Elem-type check: input.elem vs declared elem_tpe
 *      (sigma-rust coll_exists.rs:46-52: `coll.elem_tpe() != &*self.elem_tpe`)
 *      In TS, Exists MIR has no `elemTpe` field; derive from condition.args[0].tpe
 *      when condition is a FuncValue node (mirrors Filter's approach — Task 7).
 *      Throws 'coll-elem-tpe-mismatch'.
 *   4. Outer cost: addPerItemCost(3, 1, 10, n) — charges FULL length BEFORE loop
 *   5. For each item (sigma-rust coll_exists.rs:62-66):
 *      a. addCost(5)
 *      b. env.extend(argId, item)
 *      c. Eval body, assert `kind === 'Boolean'` (else 'lambda-result-type-mismatch')
 *      d. If value === true → return { kind: 'Boolean', value: true } (SHORT-CIRCUIT)
 *   6. After loop: return { kind: 'Boolean', value: false }
 *      Empty Coll → false (loop never runs; per-iter cost = 0).
 *
 * Key difference from Filter (Task 7):
 *   - Exists short-circuits on first true (Filter visits ALL items).
 *   - Exists returns Boolean; Filter returns Coll.
 *   - Outer cost params: (3, 1, 10) not (20, 1, 10).
 *
 * Key difference from Fold (Task 8):
 *   - Exists has two children (input, condition), not three.
 *   - Exists returns Boolean, not the accumulator value.
 *   - Exists short-circuits; Fold visits all items.
 *
 * Env-extend convention (same as Map/Filter — established for Tasks 7-10):
 *   TS Env is immutable. For each item we call env.extend(argId, item) to
 *   create a new scope. No save/restore needed (unlike sigma-rust's mutable env).
 *
 * Elem-type check (sigma-rust coll_exists.rs:46-52):
 *   Sigma-rust checks `coll.elem_tpe() != &*self.elem_tpe` where `self.elem_tpe`
 *   is derived from the input's SColl element type at Exists::new() construction.
 *   In TS, where Exists MIR has no `elemTpe` field, we derive the expected type
 *   from the condition when it is a FuncValue node (condition.args[0].tpe).
 *   This matches the mismatch fixture where the declared arg type (SLong) differs
 *   from the runtime input elem (SInt).
 *   When condition is not a FuncValue, skip the static check.
 *
 * Result-type check:
 *   Exists's predicate MUST return Boolean. Each per-item result is checked:
 *   if `itemRes.kind !== 'Boolean'` → throws 'lambda-result-type-mismatch'.
 *   This is explicitly enforced in TS (sigma-rust uses try_extract_into::<bool>()
 *   which also errors on non-Boolean results).
 *
 * SMOKING-GUN (outer-cost-on-full-length subtlety):
 *   Input [true, false, false, ..., false] (1000 items), predicate = identity.
 *   Short-circuit fires at item 1 (first item is true).
 *   Outer cost = addPerItemCost(3, 1, 10, 1000) = 103 (FULL n, charged BEFORE loop).
 *   Per-iter cost = 1 * 5 = 5 (only item 1 visited).
 *   Arm contribution = 108. NOT 9 (which would be outer(n=1)+5 = 4+5).
 */

import type { Exists, SValue } from '../mir/types'
import type { Env } from './env'
import type { EvalContext } from './eval-context'
import { EvalError } from './eval-context'
import { evalExpr } from './eval'
import { extractCollItems, extractFuncValue } from './_coll-helpers'
import { assertArgTypeResolved } from './_lambda'
import { sTypeEquals } from '../mir/stype-helpers'

// Outer cost: add_per_item_jit_cost(base=3, per_chunk=1, chunk_size=10, n)
// Sigma-rust ref: coll_exists.rs:60
// NOTE: base=3 — same as Fold (coll_fold.rs:48), different from Map/Filter (base=20).
const COLL_EXISTS_OUTER_BASE = 3
const COLL_EXISTS_OUTER_PER_CHUNK = 1
const COLL_EXISTS_OUTER_CHUNK_SIZE = 10

// Per-iter cost: add_jit_cost(5) per visited element before body eval
// Sigma-rust ref: coll_exists.rs:29
const COLL_EXISTS_PER_ITER = 5

/**
 * Evaluate an `Exists` node. Mixed pattern: outer after children (FULL n), per-iter in loop.
 * Short-circuits on first true; returns false for empty or no-match.
 *
 * @throws EvalError `'coll-input-not-coll'` if input does not eval to a Coll.
 * @throws EvalError `'lambda-not-callable'` if condition does not eval to a callable Lambda.
 * @throws EvalError `'coll-elem-tpe-mismatch'` if input's elem type doesn't match condition's declared arg type.
 * @throws EvalError `'lambda-result-type-mismatch'` if a body result is not Boolean.
 * @throws EvalError `'cost-limit-exceeded'` if any cost charge exceeds the limit.
 */
export function evalExists(e: Exists, env: Env, ctx: EvalContext): SValue {
  // 1. Eval input and condition — Pattern B ordering (eval children first, outer cost after).
  // Sigma-rust coll_exists.rs:18-19: input_v, condition_v in order.
  const inputVal = evalExpr(e.input, env, ctx)
  const conditionVal = evalExpr(e.condition, env, ctx)

  // 2. Guard: input must be Coll.
  const inputColl = extractCollItems(inputVal)

  // 3. Guard: condition must be a callable Lambda with at least one arg.
  const closure = extractFuncValue(conditionVal)

  // 3b. Elem-type check: input.elem must match condition's declared arg type.
  // Sigma-rust coll_exists.rs:46-52: `if coll.elem_tpe() != &*self.elem_tpe → EvalError`.
  // `self.elem_tpe` is derived from the input SColl at Exists::new() construction time.
  // TS MIR Exists has no `elemTpe` field; we derive the expected input type from
  // e.condition when it is a FuncValue MIR node (i.e., e.condition.args[0].tpe).
  // When condition is not a FuncValue node (ValUse etc.), skip — the extractFuncValue
  // guard above already enforces callable-at-runtime.
  if (e.condition.tag === 'FuncValue' && e.condition.args.length > 0) {
    const conditionInputTpe = e.condition.args[0]!.tpe
    if (!sTypeEquals(inputColl.elem, conditionInputTpe)) {
      throw new EvalError(
        `Exists: input elem type ${JSON.stringify(inputColl.elem)} does not match condition declared arg type ${JSON.stringify(conditionInputTpe)}`,
        'coll-elem-tpe-mismatch'
      )
    }
  }

  // 4. Outer cost: add_per_item_jit_cost(3, 1, 10, n) — charges FULL length BEFORE the loop.
  // Sigma-rust coll_exists.rs:60: ctx.add_per_item_jit_cost(3, 1, 10, normalized_input_vals.len())?;
  // CRITICAL: n = inputColl.items.length (FULL length, not visited count).
  // Short-circuiting at item 1 of 1000 still charges outer for n=1000.
  ctx.addPerItemCost(
    COLL_EXISTS_OUTER_BASE,
    COLL_EXISTS_OUTER_PER_CHUNK,
    COLL_EXISTS_OUTER_CHUNK_SIZE,
    inputColl.items.length
  )

  const argId = closure.argIds[0]!

  // 5. Loop: per-iter cost + env-extend + body eval + result-type check + short-circuit.
  // Sigma-rust coll_exists.rs:62-66:
  //   for item in normalized_input_vals {
  //     let res = condition_call(item)?.try_extract_into::<bool>()?;
  //     if res { return Ok(true.into()); }
  //   }
  // TS uses immutable Env.extend — no save/restore needed.
  // Short-circuit: return immediately on first true (unlike Filter which visits ALL items).
  for (const item of inputColl.items) {
    // Per-iter cost (sigma-rust coll_exists.rs:29: ctx.add_jit_cost(5)?).
    // Only charged for VISITED items — short-circuit reduces the number of charges.
    ctx.addCost(COLL_EXISTS_PER_ITER)

    // Extend the lambda's CAPTURED (definition-site) env with the arg binding
    // — lexical scoping, JVM-faithful for v6 (sigma-rust coll_exists.rs:30:
    // env.insert(func_arg.idx, arg)). For inline predicate lambdas capturedEnv
    // == the caller env (no-op); differs only for out-of-scope-captured lambdas.
    // v6 P6: reject a type-var arg type at the per-element apply (JVM
    // "Unknown type T"). Per-element ⇒ an empty input never binds, never throws.
    assertArgTypeResolved(closure.argTpes[0]!)
    const bodyEnv = closure.capturedEnv.extend(argId, item)

    // Eval body (sigma-rust coll_exists.rs:31: func_value.body.eval(env, ctx)).
    const itemRes = evalExpr(closure.body, bodyEnv, ctx)

    // Result-type check: Exists's predicate MUST return Boolean.
    // sigma-rust uses try_extract_into::<bool>() which also errors on non-Boolean.
    if (itemRes.kind !== 'Boolean') {
      throw new EvalError(
        `Exists: lambda body returned type '${itemRes.kind}' but predicate must return Boolean`,
        'lambda-result-type-mismatch'
      )
    }

    // SHORT-CIRCUIT: if predicate is true, return immediately.
    // Sigma-rust coll_exists.rs:64-66: if res { return Ok(true.into()); }
    if (itemRes.value) {
      return { kind: 'Boolean', value: true }
    }
  }

  // 6. After loop (or empty Coll): return false.
  // Sigma-rust coll_exists.rs:68: Ok(false.into())
  return { kind: 'Boolean', value: false }
}
